File: 118. Pascal's Triangle.py
#Time Complexity:O(n^2)
#Space Complexity:O(1)
class Solution(object):
    def generate(self, numRows):
        """
        :type numRows: int
        :rtype: List[List[int]]
        """
        dp=[]
        for i in range(1,numRows+1):
            dp.append([0]*i)
        for i in range(0,numRows):
            for j in range(0,i+1):
                if(j==0 or j==i):
                    dp[i][j]=1
                else:
                    dp[i][j]=dp[i-1][j-1]+dp[i-1][j]
        return dp

File: 118. Pascal's Triangle.cpp
class Solution {
public:
    /*
    Time: O(n*(n+1)/2) = O(n^2)
    Space: O(n*(n+1)/2) = O(n^2)
    */
    vector<vector<int>> generate(int r) {
        vector<vector<int>> v;

        v.push_back({1});

        for(int i=1;i<r;i++){
          v.push_back({});

          //first element
          v[i].push_back(1);

          //fill up intermediate values
          for(int j=0;j<i-1;j++){
            v[i].push_back(v[i-1][j] + v[i-1][j+1]);
          }

          //last element
          v[i].push_back(1);
        }

        return v;
        
    }
};
